### ISSUE TITLE
Stabilizing dispatch instance in Fable

### ISSUE THREAD

**Issue by vbfox** [elmish/elmish#157]
While trying to optimize react applications, passing dispatch in props of components like `PureComponent` or `React.memo` make them always update as the reference isn't currently stable (`mb.Post` results in a new call to `bind` every time a new message is received)

It can relatively easily be solved in an app by saving the dispatch instance and "stabilizing" it in the root view but that's code that every Fable use would need to use.

```fsharp
let mutable stableDispatch = None
let view model dispatch =
	let dispatch =
		match stableDispatch with
		| Some x -> x
		| None ->
			stableDispatch <- Some dispatch
			dispatch 
```

I think it should be solved in Elmish itself so it would just-work for everyone.

### Move all the code of `runWith` inside the mailbox

It has consequence on multi-thread applications as it change on what thread subscriptions are done so I don't know how compatible it is with non-fable elmish variants :

```fsharp
    let runWith (arg: 'arg) (program: Program<'arg, 'model, 'msg, 'view>) =
        let (model,cmd) = program.init arg
        MailboxProcessor.Start(fun (mb:MailboxProcessor<'msg>) ->
            let dispatch = mb.Post
            let rec loop (state:'model) =
                async {
                    let! msg = mb.Receive()
                    let newState =
                        try
                            let (model',cmd') = program.update msg state
                            program.setState model' dispatch
                            cmd' |> List.iter (fun sub -> sub dispatch)
                            model'
                        with ex ->
                            program.onError ("Unable to process a message:", ex)
                            state
                    return! loop newState
                }

            program.setState model dispatch
            let sub =
                try
                    program.subscribe model
                with ex ->
                    program.onError ("Unable to subscribe:", ex)
                    Cmd.none
            sub @ cmd |> List.iter (fun sub -> sub dispatch)

            loop model
        ) |> ignore
```

### Use a mutable variable to hold dispatch

The code is less "clean" on the elmish side but the change should be invisible to all consumers

```fsharp
    let runWith (arg: 'arg) (program: Program<'arg, 'model, 'msg, 'view>) =
        let (model,cmd) = program.init arg
        let mutable dispatch = ignore
        let inbox = MailboxProcessor.Start(fun (mb:MailboxProcessor<'msg>) ->
            let rec loop (state:'model) =
                async {
                    let! msg = mb.Receive()
                    let newState =
                        try
                            let (model',cmd') = program.update msg state
                            program.setState model' dispatch
                            cmd' |> List.iter (fun sub -> sub dispatch)
                            model'
                        with ex ->
                            program.onError ("Unable to process a message:", ex)
                            state
                    return! loop newState
                }
            dispatch <- mb.Post
            loop model
        )
        dispatch <- inbox.Post
        program.setState model dispatch
        let sub =
            try
                program.subscribe model
            with ex ->
                program.onError ("Unable to subscribe:", ex)
                Cmd.none
        sub @ cmd |> List.iter (fun sub -> sub dispatch)
```

## Conclusion

If you are ok with one of the changes, i'll send a PR

**Comment by et1975** [comment]
I thought you lost interest, but guess we're picking up from #130 👍 
I've been thinking to completely rewrite the dispatch loop and ditch the MailboxProcessor, as Fable version doesn't provide FIFO and WPF version could benefit from running on the UI thread. Let me see what I can come up with.

We still need to figure out how this can benefit event handler props that depend on the model.

**Comment by dbrattli** [comment]
Currently it is also hard to make extensions that hooks into dispatch like with [Fable.Reaction](https://github.com/dbrattli/Fable.Reaction/blob/master/src/Program.fs#L43). The only way is currently to override the view and wait for the first call, which results in very ugly code. Would it be possible to get dispatch into the program type?

**Comment by forki** [comment]
Yeah just yesterday I wanted to get the dispatch from the program to get something from outside of the normal update function into the message loop. In some way it may be considered cheating, but yes it would have helped me

**Comment by et1975** [comment]
@forki @dbrattli the subscriptions and commands returned from `init` get `dispatch` early on, but please open another issues if you want to discuss this further.

**Comment by vbfox** [comment]
@et1975 I didn't lost interest completely but I didn't manage to come to a good solution on how to handle this problem. The inversion like in #130 help to solve it but it's really hard to spread the pattern, it's also very easy to screw it at some level.

But with more reflection I realized a few things :

1. Having a stable `dispatch` instance can be enough to use PureComponent in a lot of cases. Not only it can then be used in props but passing a composed one to a sub program keeping a stable reference is doable via memoization :

	```fsharp
	let memoizeOne fn =
		let mutable cache: (_ * _) option = None
		(fun x ->
			match cache with
			| Some (key, value) when obj.ReferenceEquals(x, key) ->
				value
			| _ ->
				let v = fn x
				cache <- Some (x, v)
				v)
					  
	type SubMsg = | Hello of string
	type Msg = | Sub of SubMsg
	
	let memoizedSub: Dispatch<Msg> -> Dispatch<SubMsg> =
		memoizeOne ((>>) Sub)
	
	let subView model (dispatch: Dispatch<SubMsg>) =
		dispatch (Hello "World!")
	
	let view model (dispatch: Dispatch<Msg> -> unit) =
		subView model (memoizedSub dispatch)
	``` 
		
2. In most cases when pure components are used having a stable function passed to event handlers isn't that useful, the component should nearly never re-render and when it does it's to change the DOM, one more property mutating in this case isn't that important.
3. Specialized components like `lazyView` might be needed anyway in Elmish instead of Pure Components and they can have a different comparison function for the model (Reference or structural) and for dispatch or functions composed from dispatch (Ignored by shouldComponentUpdate)

There is also the fact that creating PureComponent require an heavy syntax for users, and doing a custom one that could be named and act as separate components in React (Something that lazyView lacks) wasn't easy. But all it changed with `React.memo` appearance making this quite possible (I'll PR the React Import repository soon and might PR the Elmish.React one too with some suggestion of utilities)

**Comment by vbfox** [comment]
On the general question of removing the `MailboxProcessor` I'm interested, I think it would be better for platforms to implement their own handling of the threading issues as they vary a lot.

**Comment by et1975** [comment]
Released #160 as 3.0.0-beta-2
